Open Policy Agent support

nickwilliams-codynamic opened the pull request:
**What changed?**

Added an `Authorizer` and `ClaimMapper` implementation to allow temporal to delegate authentication and authorization decisions to [Open Policy Agent (OPA)](https://www.openpolicyagent.org/)

**Why?**

Enables configuration of authentication and authorization requirements without having to change the temporal code base. OPA was chosen based on its open source license, deployment model, integration mechanism, and declarative language. As the implementation is simply a HTTP(S) invocation OPA could be swapped out for any HTTP API which supports the inputs and provides a valid response.

**How did you test it?**

- Unit tests created as code was written TDD
- Manually tested by running locally

**Potential risks**

If there was a bug in the configuration parser then it could prevent the service from booting, other than that I cant think of anything. The functionality needs to be opted into.

**Is hotfix candidate?**

No

nickwilliams-codynamic committed:
Add Open Policy Agent configuration

Added settings for Open Policy Agent to server
configuration file

nickwilliams-codynamic committed:
Added OPA authorizer implementation

OPA authorizer invokes an OPA data endpoint
an expects to receive a result with a boolean
allow variable.

The claims and call target are provided to the
OPA endpoint. This allows users to configure
their authorization policies externally from
temporal.

Theoretically this will work with any HTTP(s)
endpoint which accepts a POST request and returns
a 200 response with the required content

nickwilliams-codynamic committed:
Added OPA Claim mapper

OPA claim mapper simply passes through
AuthInfo via the Claim's Extensions. This
allows OPA to validate the AuthInfo as
needed.

Other AuthInfo variables, like Subject, are
left as their default values. This is to
ensure that all validation takes place
within OPA, so that temporal can be ignorant
of how validation is performed.

nickwilliams-codynamic committed:
Removed json tags from OPA authorizer structs

Removed json tags from OPA authorizer structs
as a lot of built in structs marshal using PascalCase.
The tags were set to use camelCase for the top level
properties. This caused the rego policies to look
perculiar due to the mix of casing.

This also resulted in less files being modified
for the OPA change overall.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/temporalio/temporal?pullRequest=3702) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/temporalio/temporal?pullRequest=3702) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/temporalio/temporal?pullRequest=3702) it.</sub>

nickwilliams-codynamic committed:
Fix OPA config test symlnk issues

The development-opa.yaml file was created as a symlnk
which incorrectly caused the development-cass.yaml file
to be modified.

nickwilliams-codynamic committed:
Removed use of deprecated "io/ioutil" in OPA auth

nickwilliams-codynamic committed:
Fix OPA unit tests run via buildkite

nickwilliams-codynamic committed:
Merge branch 'master' into opa-auth-support

dnr commented on config/development-opa.yaml:
```diff
@@ -0,0 +1,133 @@
+        keyspace: "temporal_visibility"
+global:
+  authorization:
+    authorizer: opa
```
Instead of adding a new development config and adding to the development docker-compose, can you move all this into the https://github.com/temporalio/samples-server repo? You can add a new directory in there with a new docker-compose, configs file, and sample policy files. (Although I think you can get away without a new config file if you add a few template options to `docker/config_template.yaml` and set env vars in the docker-compose)

nickwilliams-codynamic replied:
Sounds good, I've removed all of that configuration from this repo and will create a PR in samples-server as soon as I can.

dnr reviewed (commented):
Thanks for the contribution! I added a few comments. I don't know much about OPA so I'll trust that the protocol bits are okay.

My main concern is that I'd rather see a complete example of an OPA setup in the samples-server repo, and leave just unit tests in here, which should be enough to ensure that this client-side doesn't break.

nickwilliams-codynamic committed:
Removed dependency on OPA from unit tests

Changes the OPA unit test implementation to use
httptest rather than an OPA docker instance.

This allowed a lot of configuration to be removed
from the develop/docker-compose and
develop/buildkite directories

nickwilliams-codynamic commented:
> hanks for the contribution! I added a few comments. I don't know much about OPA so I'll trust that the protocol bits are okay.
> 
> My main concern is that I'd rather see a complete example of an OPA setup in the samples-server repo, and leave just unit tests in here, which should be enough to ensure that this client-side doesn't break.

Thanks @dnr. I completely agree with your comments, I believe I've address them with the latest commit.

With regards to the OPA protocol its quite simple and flexible. Essentially a client can POST a HTTP request with a body which looks like:

```json
{ 
  "input": {
     // Any JSON here which you want your policy to make decision using
  }
}
```

The request is sent an endpoint with the format `/v1/data/your-policy-package-here` where the policy package is determined by the policy content. For example if we had a policy in OPA which started with `package my.temporal.auth` the endpoint would be `/v1/data/my/temporal/auth`.

The response from this endpoint is of the form:

```json
{
  "result": {
        // Your policy results here
   }
}
```

So if I had a policy which looked like

```rego
package my.temporal.auth

default is_user = true
default is_admin = false
```
The JSON response would be:

```json
{
  "result": {
       "is_user": true,
       "is_admin": false
   }
}
```

Given we need Temporal to be able to understand the response I've just used the common convention of checking for a variable called `allow` which must be a boolean. This could really be any known property, but `allow` is idiomatic in the OPA world and makes sense generally so I thought it was good choice.

So from a Temporal view this really just a webhook which sends the following JSON to any HTTP endpoint you want to configure
```json
{
   "input": {
      "Claims": {
          // The claims
      },
      "Target": {
          // The target
      }
   }
}
```

and requires a 200 response with the content
```json
{
   "result": {
       "allow": true
    }
}
```

nickwilliams-codynamic committed:
Removed unused struct from OPA authorizer tests

nickwilliams-codynamic committed:
Removed config tests related to OPA

nickwilliams-codynamic committed:
Merge branch 'master' into opa-auth-support

nickwilliams-codynamic committed:
Merge branch 'master' into opa-auth-support

nickwilliams-codynamic commented:
Hi @dnr  is there anything else you wanted changing with this PR?

dnr commented:
Hi, really sorry about the delay on this one.

In general I think the code is great, there's two general reasons I haven't merged it yet:

First, we don't have a great system right now for this category of "contributed" code that lives in this repo but isn't quite supported as part of the core product. Other Temporal projects, e.g. the Go SDK, have a `contrib` directory where this sort of thing can live, but there isn't one for the server yet. I'd like to have a discussion with the team first about either setting that up, or taking on the responsibility for supporting this piece of code. (Which is admittedly pretty simple, but there's an ongoing testing burden to keep it working as we make changes.)

Second, after spending some time with the "default" authorizer this month, I have a slightly different opinion about how these pieces should fit together: it seems to me that the default authorizer implements the "obviously correct" logic for allowing and denying calls based on claims and target (as of version 1.20). Specifically, it has some knowledge about the semantics of different methods (read-only vs read-write; requires namespace vs not) that we'll keep up to date. Using this authorizer, that logic would have to be replicated in an OPA config, which seems fragile.

I'm thinking that the right place for pluggable logic is the claim mapper: it should call out to OPA to get claims, and then the default authorizer can handle allowing/denying based on the target and claims. This is just an opinion, though :slightly_smiling_face:.  If you have some authorization logic that couldn't fit in that structure, I'd be curious to know what it is.

dnr commented on common/authorization/opa_claim_mapper.go:
```diff
@@ -0,0 +1,51 @@
+	}
+
+	opaClaimExtensions struct {
+		AuthInfo AuthInfo `json:"authInfo"`
```
The TLS fields in here seem like they could get pretty big. Does it really make sense to pass full client cert chains (keeping in mind that they've already been validated here)? If not, should we maybe copy this struct and clear some of the larger and less useful fields?

`TLSSubject.Names` is probably not useful since the common ones are parsed into the fields above it (and of course everyone only uses CommonName)

`TLSConnection.State.PeerCertificates`, `VerifiedChain`, etc. could be quite large and probably were not intended to be encoded as JSON

Do you have an example of what a typical value looks like in JSON for TLS and non-TLS?

nickwilliams-codynamic replied:
Good point, I'll have a look into this. So long as we provide the FQDN I guess that should be enough for most cases, maybe we should include the fingerprint too?

dnr commented on common/config/config.go:
```diff
@@ -550,6 +550,8 @@ type (
 		// Empty string for noopClaimMapper or "default" for defaultJWTClaimMapper
 		ClaimMapper string `yaml:"claimMapper"`
+		// Empty string unless using "opa" Authorizer. The HTTP(s) endpoint which temporal will invoke for authorization decisions.
+		OpaEndpoint string `yaml:"opaEndpoint"`
```
I'd like to make a new `AuthorizerConfig any` field here so authorizers/claim mappers can define their own config, but I might refactor that later

dnr commented on common/authorization/authorizer.go:
```diff
@@ -87,6 +87,8 @@ func GetAuthorizerFromConfig(config *config.Authorization) (Authorizer, error) {
 		return NewNoopAuthorizer(), nil
 	case "default":
 		return NewDefaultAuthorizer(), nil
+	case "opa":
```
There's nothing really specific to OPA at all here, is there? Should we call it a `http` authorizer?

nickwilliams-codynamic replied:
Yep HTTP makes more sense. I will update that.

dnr reviewed (commented):
Just a few more small comments

nickwilliams-codynamic commented:
> I'm thinking that the right place for pluggable logic is the claim mapper: it should call out to OPA to get claims, and then the default authorizer can handle allowing/denying based on the target and claims. This is just an opinion, though 🙂. If you have some authorization logic that couldn't fit in that structure, I'd be curious to know what it is.

I think that moving this to the claim mapper would be okay. It might be I was trying to go too fine grained or that I misunderstood the default authorizer.  One thing I'd be keen to be able to support was allowing some client to signal a workflow in namespace A, but not attach a worker to a task queue in namespace A. As I understood it they are both `write` operations so can't be split apart by the default authorizer. For example:

- We use namespaces to isolate teams workloads
- If we image there are two namespaces `team-1` and `team-2` 
- Team 1 deploy workers associated with the `team-1` namespace, the worker can generate a JWT with the `team-1:owner` claim so can poll for tasks from the `team-1` namespace
- Team 2 deploy workers associated with the `team-2` namespace, the worker can generate a JWT with the `team-2:owner` and `team-1:consumer` claim so can poll for tasks from the `team-2` namespace and signal workflows in the `team-1` namespace

If this could be done in the claims mapper then that is great, but I couldn't see how, as the `read` / `write` buckets of endpoints are quite big. We we're hoping to ensure that a team couldn't pull work from another team's namespace, either by accident or maliciously. We can prevent access to the data using encryption converters I suppose, but it would be nice to prevent the API calls themselves.

nwilliams1-bigw commented:
@dnr In addition to the comment above I don't think the default authorizer is good enough for a more general use case which I was hoping this HTTP authorizer would be able to support.  One problem I can see at the moment:

Only users with `system:read` claim can access the Temporal UI - as the UI invokes `globalReadonlyAPIs`. This means team members who have access to a single namespace can't use the UI or they will have access to read every namespaces workflows (via system:read)

For our use case we would want to allow read access to `/namespaces`, `/search-attributes`, and `/cluster` to anyone with a `{namespace}:read` claim so that the UI functions, but deny access to namespace-specific APIs. This isn't possible with the default authorizer.

Would you be okay with me proceeding with this http implementation at the authorizer level.

nickwilliams-codynamic committed:
Merge branch 'master' of https://github.com/temporalio/temporal into opa-auth-support

nickwilliams-codynamic committed:
Rename OPA auth to HTTP auth

nickwilliams-codynamic committed:
Remove detailed TLSInfo from serialized message for HTTP auth

nickwilliams-codynamic committed:
Fix typo in description of config.go for HTTP auth

mbernier commented:
For vis - There's a related proposal here: https://github.com/temporalio/proposals/issues/81

fakeNews-jpg commented:
Hello @MichaelSnowden, any update on this ? Would be an amazing feature.. :)

barryib commented:
Hey, this implementation is great. Is there a way to move forward with this PR ? Or maybe we should ship it as an external authorizer plugin instead in a different repo ?

nickwilliams-codynamic committed:
Merge branch 'main' into opa-auth-support

davidspek commented:
Any chance in this being picked up again?
